### PR TITLE
Test all storage modes

### LINKED ISSUE #312: Reductions don't work on Shared Arrays
```julia
using Metal

a = mtl(rand(10_240_000), storage=Shared)
```

```sh
julia> sum(a)
0.0f0`
```

The issue might be caused by #305 

==> missing sync in Base.unsafe_convert(::Type{Ptr{S}}, x::MtlArray{T}) where {S, T}

### PR DESCRIPTION
**Pull request by christiangnrd**
Addresses https://github.com/JuliaGPU/Metal.jl/issues/312#issuecomment-1988304979 and closes #312.

### CODE REVIEW

File: src/array.jl [hunk 1]
```diff
@@ -170,14 +181,16 @@ end
 
 function Base.unsafe_convert(::Type{MtlPointer{T}}, x::MtlArray) where {T}
    buf = x.data[]
+   synchronize()
```
Comment (tgymnich) [inline src/array.jl:184]: I don’t think we need to synchronize here. MtlPointers will only be used on device.

**Review by tgymnich — COMMENTED** [review]
> Addresses https://github.com/JuliaGPU/Metal.jl/issues/312#issuecomment-1988304979 but does not fix https://github.com/JuliaGPU/Metal.jl/issues/312.

it does seem to me like your changes would fix #312

File: src/array.jl (hunk 1 above)
Comment (christiangnrd) [reply src/array.jl:184]: Without this the kernelabstractions `copyto` test fails.

**Comment by christiangnrd** [comment]
> it does seem to me like your changes would fix #312

I commented that before I decided to submit the fixes. The `gpuarrays/indexing scalar` tests are still failing with Shared and Managed storage modes.

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -29,16 +29,40 @@ steps:
         matrix:
           setup:
             julia:
+              - "1"
```
Comment (maleadt) [inline .buildkite/pipeline.yml:32]: Why this change? I prefer an explicit enumeration; it's unlikely that GPU back-ends will happen to support the latest Julia release if we haven't explicitly added support.

Comment (christiangnrd) [reply .buildkite/pipeline.yml:32]: I was in the process of undoing it I don't like it either.

**Comment by christiangnrd** [comment]
I think everything in this PR is ready to go except isn't it a bit of a code smell for there to be a `synchronize()` in an unsafe function?

File: src/array.jl (hunk 1 above)
Comment (tgymnich) [reply src/array.jl:184]: Before #305 it worked without a sync. https://github.com/JuliaGPU/Metal.jl/blob/91d72d0e199a950ce9f23912375c0ccbf6c0e337/src/array.jl#L147

Comment (tgymnich) [reply src/array.jl:184]: So the problem is, that pointer(A) might mean device or host pointer. https://github.com/JuliaGPU/Metal.jl/pull/316 might fix that.

Comment (tgymnich) [reply src/array.jl:184]: Here is what CUDA.jl does https://github.com/JuliaGPU/CUDA.jl/blob/bb49887198f258ffcb186d81df4a787453428b38/src/CUDAKernels.jl#L41

Comment (christiangnrd) [reply src/array.jl:184]: Since this PR passes all tests, I suggest we merge this and then you can finish up the fix in #316.

Comment (maleadt) [reply src/array.jl:184]: FWIW, it's really not acceptable to sync on every pointer conversion, as that would remove almost all opportunity for asynchronous execution (as most GPU operations involve a conversion to a pointer).
